src/js/reducers/datasetsReducer.js

```diff
@@ -33,8 +33,23 @@ function datasetsReducer(state, action) {
       field: action.field,
       order: action.order
     }));
+  }
+
+  if (action.type === ACTIONS.FILTER_DATASET) {
+    state = setIn(state, action.id + '._filter',
+    Immutable.fromJS({
+      expression: action.expression
+    }));
+  }
 
+  if (action.type === ACTIONS.SHOW_EXPRESSION_TEXTBOX) {
+    state = setIn(state, action.id + '._expressionTextbox',
+    Immutable.fromJS({
+      show: action.show,
+      time: action.time
+    }));
```

Comment:
I'm not sure this logic should be part of the store. Instead, because it's so specific to a particular React component, this may be one of the rare cases that we want to use the internal state of the component.